### PR TITLE
MetaTensorDesc changes for reconstructing proper FakeTensors

### PR DESCRIPTION
**Pull request by aorenste**
A few changes to MetaTensorDesc and friends:

1. Change view_func from a raw method to an ADT where the common case (FakeTensor._view_func_unsafe) is a simple representation instead.
2. (minor) Remove and fix some `type: ignore`s added by #141839
3. (minor) Fix _UNSERIALIZABLE to be a set instead of a dict which is converted into a set each time it's used.
 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #142072
* #141659
* __->__ #141926

### CODE REVIEW

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -345,9 +356,11 @@ def describe_tensor(
 
         from torch.nested._internal.nested_tensor import _tensor_symint_registry
 
+        view_func = _ViewFunc.from_tensor(t)
+
         # TODO: Is it important to enable torch.inference_mode before querying
         # these values?
-        r = MetaTensorDesc(
```
Comment (aorenste) [inline torch/_subclasses/meta_utils.py:350]: mypy is being weird and requiring a type decl here...

File: torch/_subclasses/meta_utils.py [hunk 2]
```diff
@@ -476,6 +487,64 @@ def as_json(self, describer_id: _DescriberId) -> Dict[str, object]:
         }
 
 
+@dataclass(frozen=True)
+class _ViewFunc(Generic[_TensorT]):
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:493]: FWIW I don't really see why this should be private, if MetaTensorDesc isn't considered private per this file.

File: torch/_subclasses/meta_utils.py [hunk 3]
```diff
@@ -1582,16 +1643,14 @@ def is_c_of_r(
                                 sizes,
                                 strides,
                                 dtype=t.dtype,
-                                device="meta",
+                                device=override_device or t.device,
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py]: I can't tell if this is semantics preserving or not

File: torch/_subclasses/meta_utils.py [hunk 4]
```diff
@@ -1809,6 +1865,7 @@ def __call__(
                 callback_,
                 source,
                 symbolic_context,
+                override_device="meta",
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py]: oh I guess it is

**Comment by ezyang** [comment]
I think you need to recursively propagate the force device thing. But I am also not too sure about what that is about

File: torch/_subclasses/meta_utils.py (hunk 3 above)
Comment (aorenste) [reply torch/_subclasses/meta_utils.py]: This is on the older version of this PR - there was a collision w/ #141839 (which landed yesterday) which implements similar functionality in a different way - so I ripped the override stuff out and just used that.

File: torch/_subclasses/meta_utils.py (hunk 2 above)
Comment (aorenste) [reply torch/_subclasses/meta_utils.py:493]: It doesn't have to be - but it's kind of an implementation detail...

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**